mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala

```diff
@@ -204,17 +204,54 @@ class KMeans private (
         + " parent RDDs are also uncached.")
     }
 
-    // Compute squared norms and cache them.
-    val norms = data.map(Vectors.norm(_, 2.0))
-    norms.persist()
-    val zippedData = data.zip(norms).map { case (v, norm) =>
-      new VectorWithNorm(v, norm)
+    val zippedData = data.map { x =>
+      val norm = Vectors.norm(x, 2.0)
+      new VectorWithNorm(x, norm)
     }
-    val model = runAlgorithm(zippedData)
-    norms.unpersist()
+
+    val centers = initialModel match {
+      case Some(kMeansCenters) => {
+        kMeansCenters.clusterCenters.map(s => new VectorWithNorm(s))
+      }
+      case None => {
+        if (initializationMode == KMeans.RANDOM) {
+          initRandom(zippedData)
+        } else {
+          initKMeansParallel(zippedData)
+        }
+      }
+    }
+
+    val samplePoint = data.first()
+    val dims = samplePoint.size
+    // TODO: make stack size can be configured.
+    val stackSize = 128
+
+    val blockData = zippedData.mapPartitions { iter =>
+      iter.grouped(stackSize).map { points =>
+        val realSize = points.size
+        val pointsArray = new Array[Double](realSize * dims)
+        val pointsNormArray = new Array[Double](realSize)
+        var numRows = 0
+
+        points.foreach { point =>
+          System.arraycopy(point.vector.toArray, 0, pointsArray, numRows * dims, dims)
+          pointsNormArray(numRows) = math.pow(point.norm, 2.0)
+          numRows += 1
+        }
+        val pointMatrix = new DenseMatrix(numRows, dims, pointsArray, true)
+        val pointsNormMatrix = new DenseMatrix(numRows, k, Array.fill(k)(pointsNormArray).flatten)
```

Comment:
Collection methods might be expensive in that case, because `fill` will create array of arrays, copy data there, and then `flatten` will create another array and copy data there. Copy the array directly with `System.arraycopy` using additional loop instead.